### PR TITLE
Taimoor's Migrations and Associations

### PR DESCRIPTION
**Pull request by taimoormalik97**

### CODE REVIEW

File: app/models/feedback.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class Feedback < ApplicationRecord
+  belongs_to :doctor
```
Comment (sulemanuzair) [inline app/models/feedback.rb:3]: who will provide feedback and on what basis ?
do we need to store the information who provided feedback ?

File: db/migrate/20200908082727_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      t.string :gender
+      t.date :dob
+      t.text :family_history
+      t.string :registration_no
+      t.string :speciality
+      t.decimal :consultancy_fee, precision: 10, scale: 2
+
+      t.timestamps
+    end
+    add_index :users, :name
```
Comment (sulemanuzair) [inline db/migrate/20200908082727_create_users.rb]: why adding index on name ?

File: app/models/feedback.rb (hunk 1 above)
Comment (taimoormalik97) [reply app/models/feedback.rb:3]: actually we thought that feedback should remain anonymous , and patient's information shouldn't be stored here

File: db/migrate/20200908082727_create_users.rb (hunk 2 above)
Comment (taimoormalik97) [reply db/migrate/20200908082727_create_users.rb]: i will remove the index, do i also need to add null: false with all columns?

File: test/fixtures/admins.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (sulemanuzair) [inline test/fixtures/admins.yml]: remove unused files (.yml and empty test cases)

Comment (taimoormalik97) [reply test/fixtures/admins.yml]: actually those files are part of webpacker

File: app/models/admin.rb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+class Admin < User
+  has_many :purchase_orders, dependent: :destroy
```
Comment (sulemanuzair) [inline app/models/admin.rb]: we do not want the history of purchase orders to be deleted if we allow admin to be deleted

File: app/models/admin.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class Admin < User
```
Comment (sulemanuzair) [inline app/models/admin.rb:1]: add belongs to Hospital in all models and migrations, if defined in parent class, then not required in child classes

File: app/models/patient.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class Patient < User
+  has_many :appointments, dependent: :destroy
```
Comment (sulemanuzair) [inline app/models/patient.rb]: if we delete a doctor, past appointment of patients would be deleted
and vice versa

Comment (sulemanuzair) [reply app/models/patient.rb]: needs discussion

File: db/migrate/20200908082727_create_users.rb (hunk 2 above)
Comment (sulemanuzair) [reply db/migrate/20200908082727_create_users.rb]: also add composite index on id and type

File: app/models/doctor.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class Doctor < User
+  has_many :appointments, dependent: :destroy
+  has_many :patients, through: :appointments
+  has_many :feedbacks, dependent: :destroy
+  has_many :availabilities, dependent: :destroy
+  has_many :bill_details as: :billable, dependent: :destroy
```
Comment (sulemanuzair) [inline app/models/doctor.rb]: needs discussion, why doctor has bill_details and bills

File: app/models/admin.rb (hunk 4 above)
Comment (taimoormalik97) [reply app/models/admin.rb]: ok updated

File: app/models/admin.rb (hunk 5 above)
Comment (taimoormalik97) [reply app/models/admin.rb:1]: done

File: db/migrate/20200908082727_create_users.rb (hunk 2 above)
Comment (sulemanuzair) [reply db/migrate/20200908082727_create_users.rb]: yes

File: test/fixtures/admins.yml (hunk 3 above)
Comment (sulemanuzair) [reply test/fixtures/admins.yml]: files having path of test/ spec/ are part of test cases, remove them in this PR, if they are required in the rails application, create separate PR for it

File: app/models/doctor.rb (hunk 7 above)
Comment (taimoormalik97) [reply app/models/doctor.rb]: because here doctors bill will also be generated when a patient visits him

File: test/fixtures/admins.yml (hunk 3 above)
Comment (taimoormalik97) [reply test/fixtures/admins.yml]: done

File: app/models/feedback.rb (hunk 1 above)
Comment (sulemanuzair) [reply app/models/feedback.rb:3]: include feedback based on appointment, 
if we want the patient to remain anonymous, we can hide its information on view

File: db/migrate/20200908124238_create_hospitals.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+class CreateHospitals < ActiveRecord::Migration[6.0]
+  def change
+    create_table :hospitals do |t|
+      t.string :name
+      t.string :address
+      t.bigint :phone_number
+
+      t.timestamps
```
Comment (sulemanuzair) [inline db/migrate/20200908124238_create_hospitals.rb]: null: false

File: test/models/admin_test.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (sulemanuzair) [inline test/models/admin_test.rb]: remove test files

File: test/models/admin_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+
+class AdminTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
```
Comment (sulemanuzair) [inline test/models/admin_test.rb]: add indices on hospital_id as well in all tables

**Comment by sulemanuzair** [comment]
cycle1

**Pull request merged**